src/main/scala/firrtl/transforms/ConstantPropagation.scala

```diff
@@ -420,30 +444,36 @@ class ConstantPropagation extends Transform with ResolvedAnnotationPaths {
-            case lit: Literal => Some(pad(lit, ltpe))
-            case WRef(regName, _, RegKind, _) if (regName == lname) => Some(e)
-            case WRef(nodeName, _, NodeKind, _) => nodeMap.get(nodeName).flatMap(regConstant(_))
-            case Mux(_, tval, fval, _) => (regConstant(tval), regConstant(fval)) match {
-              case (Some(wr: WRef), Some(x)) if weq(lref, wr) => Some(x) // Mux(_, selfassign, <constRHSCandidate>)
-              case (Some(x), Some(wr: WRef)) if weq(lref, wr) => Some(x) // Mux(_, <constRHSCandidate>, selfassign)
-              case (x, y) if (x == y) => x                               // No-op mux
-              case _ => None                                             // At least one case not constant-convertible
-            }
-            case _ => None
+         /* Checks if an RHS expression e of a register assignment is convertible to a constant assignment.
+          * Here, this means that e must be 1) a literal, 2) a self-connect, or 3) a mux tree of
+          * cases (1) and (2).  In case (3), it also recursively checks that the two mux cases can
+          * be resolved: each side is allowed one candidate register and one candidate literal to
+          * appear in their source trees, referring to the potential constant propagation case that
+          * they could allow. If the two are compatible (no different bound sources of either of
+          * the two types), they can be resolved by combining sources. Otherwise, they propagate
+          * NonConstant values.  When encountering a node reference, it expands the node by to its
+          * RHS assignment and recurses.
+          *
+          * @return a RegCPEntry describing the constant prop-compatible sources driving this expression
+          */
+          def regConstant(e: Expression): RegCPEntry = e match {
+            case lit: Literal => RegCPEntry(UnboundConstant, BoundConstant(lit))
+            case WRef(regName, _, RegKind, _) => RegCPEntry(BoundConstant(regName), UnboundConstant)
+            case WRef(nodeName, _, NodeKind, _) if nodeMap.contains(nodeName) =>
+              nodeRegCPEntries.getOrElseUpdate(nodeName, { regConstant(nodeMap(nodeName)) })
+            case Mux(_, tval, fval, _) => regConstant(tval).resolve(regConstant(fval))
+            case _ => RegCPEntry(NonConstant, NonConstant)
           }
-          def cpExp(e: Expression) = constPropExpression(nodeMap, instMap, constSubOutputs)(e)
-          regConstant(rhs).foreach {
-            case wr: WRef => nodeMap(lname) = cpExp(pad(passes.RemoveValidIf.getGroundZero(ltpe), ltpe))
-            case e => nodeMap(lname) = cpExp(e)
+          val zero = passes.RemoveValidIf.getGroundZero(ltpe)
+          def padCPExp(e: Expression) = constPropExpression(nodeMap, instMap, constSubOutputs)(pad(e, ltpe))
+          regConstant(rhs) match {
+            case RegCPEntry(BoundConstant(name), litBinding) if (name == lname) => litBinding match {
```

Comment:
```suggestion
            case RegCPEntry(BoundConstant(`lname`), litBinding) => litBinding match {
```
I prefer this to using following if guards but doesn't really matter.

Reply:
Haha, that's how I originally wrote it, but I changed it since there is so little backtick pattern match usage in firrtl, and I thought it might be confusing.